### PR TITLE
feat: Deno.assert namespace

### PR DESCRIPTION
**Pull request by bartlomieju**
This PR:
* adds `Deno.asserts` namespace which provides all methods available in `std/testing/asserts.ts` in a single namespace (`assert` === `Deno.assert`, `assertEquals` === `Deno.asserts.equals`, etc.)
* adds `Deno.errors.AssertionError`
* uses formatting provided by `Deno.inspect()` for assertion messages instead of bringing `std/testing/format.ts` along
* does not provide diffing in output for `AssertionError`

I'm completely open to make scope of this PR smaller/greater if we decide so.

Ref #4090 

CC @ry @kitsonk

### CODE REVIEW

File: cli/js/lib.deno.ns.d.ts [hunk 1]
```diff
@@ -17,6 +17,68 @@ declare namespace Deno {
...
+    (expr: unknown, msg?: string): asserts expr;
+    /** Use this to stub out methods that will throw when invoked. */
+    unimplemented(msg?: string): never;
+    /** Use this to assert unreachable code. */
+    unreachable(msg?: string): never;
+    /**
+     * Make an assertion that `actual` and `expected` are equal, deeply. If not
+     * deeply equal, then throw.
+     */
+    equals(actual: unknown, expected: unknown, msg?: string): void;
```
Comment (bartlomieju) [inline cli/js/lib.deno.ns.d.ts]: As suggested by @kitsonk in #4090 collapsed into a single namespace

File: cli/js/lib.deno.ns.d.ts [hunk 2]
```diff
@@ -17,6 +17,68 @@ declare namespace Deno {
     name: string;
   }
 
+  interface Constructor {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    new (...args: any[]): any;
+  }
+
+  export interface Assert {
```
Comment (bartlomieju) [inline cli/js/lib.deno.ns.d.ts]: Probably needs to be marked as `UNSTABLE: New API`

File: cli/js/lib.deno.ns.d.ts [hunk 3]
```diff
@@ -17,6 +17,68 @@ declare namespace Deno {
...
+    /** Make an assertion, if not `true`, then throw. */
+    (expr: unknown, msg?: string): asserts expr;
+    /** Use this to stub out methods that will throw when invoked. */
+    unimplemented(msg?: string): never;
```
Comment (nayeemrmn) [inline cli/js/lib.deno.ns.d.ts]: This one should be removed. It's not actually an assertion and wouldn't be used in tests but the implementations -- https://github.com/denoland/deno/issues/4052#issuecomment-589254004 would apply. It also applies to `unreachable()` but that one is conceptually more for tests which are already non-portable so it maybe gets a pass.

File: cli/js/lib.deno.ns.d.ts [hunk 4]
```diff
@@ -1315,6 +1373,7 @@ declare namespace Deno {
     UnexpectedEof: ErrorConstructor;
     BadResource: ErrorConstructor;
     Http: ErrorConstructor;
+    AssertionError: ErrorConstructor;
```
Comment (bartlomieju) [inline cli/js/lib.deno.ns.d.ts]: This feels wrong, there will be type information about `actual` and `expected` 🤔

**Comment by bartlomieju** [comment]
I've been pondering for a few days and I think this PR should be scaled back significantly and start with only `assert` and `AssertionError`. 

```ts
export class AssertionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AssertionError";
  }
}

export function assert(expr: unknown, msg = ""): asserts expr {
  if (!expr) {
    throw new AssertionError(msg);
  }
}
```

Because of the signature of `assert` it is very useful when using writing user code, and I'm not even talking about tests, but actual scripts. Consider example below:
```ts
const p = Deno.run({ 
  args: ["deno", "--help"], 
  stdout: "piped"
});

// TS complains; `p.stdout` might be undefined
const output = Deno.readAll(p.stdout);

// TS doesn't complain
Deno.assert(p.stdout);
const output = Deno.readAll(p.stdout);
```

**Comment by ry** [comment]
I would be much happier if this was scaled back. But if it’s just about adding Deno.assert why not just use console.assert?

**Comment by bartlomieju** [comment]
> I would be much happier if this was scaled back. But if it’s just about adding Deno.assert why not just use console.assert?

`console.assert` just gives you message to stderr if assertion failed - there's really not way to try/catch it and do further processing.

**Comment by bartlomieju** [comment]
Closing because stale, we'll rediscuss this issue soon

**Comment by jakajancar** [comment]
It would be nice if it was specified if `std/testing/asserts.ts` is OK to use in production code as opposed to just in tests.

Also, unrelated, I'm curious what is the criteria for something going into `Deno` vs std.

**Pull request closed without merging**